File: Collections/MultiDictionaryList.vb
Public Structure MultiDictionaryList(Of TKey, TValue)
    Implements IList(Of TValue),
               IList

    Public Sub New(ByVal Parent As MultiDictionary(Of TKey, TValue), ByVal Key As TKey, ByVal List As List(Of TValue))
        Me._Parent = Parent
        Me._Key = Key
        Me.List = List
        Me.Version = Me.Parent.Version
    End Sub

    Private Sub CheckChanges()
        If Me.Version <> Me.Parent.Version Then
            Dim T As List(Of TValue) = Nothing
            If Me.Parent.Dic.TryGetValue(Me.Key, T) Then
                Me.List = T
            Else
                Me.List = Nothing
            End If
            Me.Version = Me.Parent.Version
        End If
    End Sub

    Private Sub BeforeInp()
        Me.CheckChanges()
        If Me.List Is Nothing Then
            Me.List = New List(Of TValue)()
            Me.Parent.ReportKeyFilled(Me)
        End If
    End Sub

    Private Sub AfterOut()
        Me.CheckChanges()
        If Me.List.Count = 0 Then
            Me.List = Nothing
            Me.Parent.ReportKeyEmpty(Me)
        End If
    End Sub

    Public Sub Add(item As TValue) Implements ICollection(Of TValue).Add
        Me.CheckChanges()
        Me.BeforeInp()
        Me.List.Add(item)
    End Sub

    Public Sub Insert(index As Integer, item As TValue) Implements IList(Of TValue).Insert
        Me.CheckChanges()
        Me.BeforeInp()
        Me.List.Insert(index, item)
    End Sub

    Public Sub Clear() Implements ICollection(Of TValue).Clear, IList.Clear
        Me.CheckChanges()
        If Me.List IsNot Nothing Then
            Me.List.Clear()
            Me.AfterOut()
        End If
    End Sub

    Public Sub RemoveAt(index As Integer) Implements IList(Of TValue).RemoveAt, IList.RemoveAt
        Me.CheckChanges()
        If Me.List Is Nothing Then
            Throw New ArgumentOutOfRangeException("index")
        End If
        Me.List.RemoveAt(index)
        Me.AfterOut()
    End Sub

    Public Function Remove(item As TValue) As Boolean Implements ICollection(Of TValue).Remove
        Me.CheckChanges()
        If Me.List Is Nothing Then
            Return False
        End If
        Dim R = Me.List.Remove(item)
        Me.AfterOut()
        Return R
    End Function

    Public Function Contains(item As TValue) As Boolean Implements ICollection(Of TValue).Contains
        Me.CheckChanges()
        If Me.List Is Nothing Then
            Return False
        End If
        Return Me.List.Contains(item)
    End Function

    Public Function IndexOf(item As TValue) As Integer Implements IList(Of TValue).IndexOf
        Me.CheckChanges()
        If Me.List Is Nothing Then
            Return -1
        End If
        Return If(Me.List, EmptyList).IndexOf(item)
    End Function

    Public Sub CopyTo(array() As TValue, arrayIndex As Integer) Implements ICollection(Of TValue).CopyTo
        Me.CheckChanges()
        If Me.List IsNot Nothing Then
            Me.List.CopyTo(array, arrayIndex)
        End If
    End Sub

    Private Sub IList_CopyTo(array As Array, index As Integer) Implements ICollection.CopyTo
        Me.CheckChanges()
        If Me.List IsNot Nothing Then
            DirectCast(Me.List, IList).CopyTo(array, index)
        End If
    End Sub

    Public Function GetEnumerator() As List(Of TValue).Enumerator
        Me.CheckChanges()
        Return If(Me.List, EmptyList).GetEnumerator()
    End Function

    Default Public Property Item(index As Integer) As TValue Implements IList(Of TValue).Item
        Get
            Me.CheckChanges()
            If Me.List Is Nothing Then
                Throw New ArgumentOutOfRangeException("index")
            End If
            Return Me.List.Item(index)
        End Get
        Set(value As TValue)
            Me.CheckChanges()
            If Me.List Is Nothing Then
                Throw New ArgumentOutOfRangeException("index")
            End If
            Me.List.Item(index) = value
        End Set
    End Property

    Public ReadOnly Property Count As Integer Implements ICollection(Of TValue).Count, ICollection.Count
        Get
            Me.CheckChanges()
            If Me.List Is Nothing Then
                Return 0
            End If
            Return Me.List.Count
        End Get
    End Property


#Region "Obvious Implementations"
    ' The following methods call the ones from above.

    Private Function IList_Contains(value As Object) As Boolean Implements IList.Contains
        Return Me.Contains(DirectCast(value, TValue))
    End Function

    Private Function IList_IndexOf(value As Object) As Integer Implements IList.IndexOf
        Return Me.IndexOf(DirectCast(value, TValue))
    End Function

    Private Function IEnumerable_1_GetEnumerator() As IEnumerator(Of TValue) Implements IEnumerable(Of TValue).GetEnumerator
        Return Me.GetEnumerator()
    End Function

    Private Function IEnumerable_GetEnumerator() As IEnumerator Implements IEnumerable.GetEnumerator
        Return Me.GetEnumerator()
    End Function

    Private Function IList_Add(value As Object) As Integer Implements IList.Add
        Me.Add(DirectCast(value, TValue))
        Return Me.Count - 1
    End Function

    Private Sub IList_Insert(index As Integer, value As Object) Implements IList.Insert
        Me.Insert(index, DirectCast(value, TValue))
    End Sub

    Private Sub IList_Remove(value As Object) Implements IList.Remove
        Me.Remove(DirectCast(value, TValue))
    End Sub

    Private Property IList_Item(index As Integer) As Object Implements IList.Item
        Get
            Return Me.Item(index)
        End Get
        Set(value As Object)
            Me.Item(index) = DirectCast(value, TValue)
        End Set
    End Property

    Private ReadOnly Property IsReadOnly As Boolean Implements ICollection(Of TValue).IsReadOnly, IList.IsReadOnly
        Get
            Return False
        End Get
    End Property

    Private ReadOnly Property IsFixedSize As Boolean Implements IList.IsFixedSize
        Get
            Return False
        End Get
    End Property

    Private ReadOnly Property SyncRoot As Object Implements ICollection.SyncRoot
        Get
            Throw New NotSupportedException()
        End Get
    End Property

    Private ReadOnly Property IsSynchronized As Boolean Implements ICollection.IsSynchronized
        Get
            Return False
        End Get
    End Property
#End Region

#Region "Parent Property"
    Private ReadOnly _Parent As MultiDictionary(Of TKey, TValue)

    Public ReadOnly Property Parent As MultiDictionary(Of TKey, TValue)
        Get
            Return Me._Parent
        End Get
    End Property
#End Region

#Region "Key Property"
    Private ReadOnly _Key As TKey

    Public ReadOnly Property Key As TKey
        Get
            Return Me._Key
        End Get
    End Property
#End Region

    Private Version As Byte
    Friend List As List(Of TValue)
    Private Shared ReadOnly EmptyList As List(Of TValue) = New List(Of TValue)()

End Structure

File: Collections/MultiDictionaryList.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace Ks.Common
{
    public struct MultiDictionaryList<TKey, TValue> : IList<TValue>, IList
    {
        public MultiDictionaryList(MultiDictionary<TKey, TValue> Parent, TKey Key, List<TValue> List)
        {
            this.Parent = Parent;
            this.Key = Key;
            this.List = List;
            this.Version = Parent.Version;
        }

        private void CheckChanges()
        {
            if (this.Version != this.Parent.Version)
            {
                if (this.Parent.Dic.TryGetValue(this.Key, out var T))
                {
                    this.List = T;
                }
                else
                {
                    this.List = null;
                }

                this.Version = this.Parent.Version;
            }
        }

        private void BeforeInp()
        {
            this.CheckChanges();
            if (this.List == null)
            {
                this.List = new List<TValue>();
                this.Parent.ReportKeyFilled(this);
            }
        }

        private void AfterOut()
        {
            this.CheckChanges();
            if (this.List.Count == 0)
            {
                this.List = null;
                this.Parent.ReportKeyEmpty(this);
            }
        }

        public void Add(TValue item)
        {
            this.CheckChanges();
            this.BeforeInp();
            this.List.Add(item);
        }

        public void Insert(int index, TValue item)
        {
            this.CheckChanges();
            this.BeforeInp();
            this.List.Insert(index, item);
        }

        public void Clear()
        {
            this.CheckChanges();
            if (this.List != null)
            {
                this.List.Clear();
                this.AfterOut();
            }
        }

        public void RemoveAt(int index)
        {
            this.CheckChanges();
            if (this.List == null)
            {
                throw new ArgumentOutOfRangeException(nameof(index));
            }

            this.List.RemoveAt(index);
            this.AfterOut();
        }

        public bool Remove(TValue item)
        {
            this.CheckChanges();
            if (this.List == null)
            {
                return false;
            }

            var R = this.List.Remove(item);
            this.AfterOut();
            return R;
        }

        public bool Contains(TValue item)
        {
            this.CheckChanges();
            if (this.List == null)
            {
                return false;
            }

            return this.List.Contains(item);
        }

        public int IndexOf(TValue item)
        {
            this.CheckChanges();
            return (this.List ?? EmptyList).IndexOf(item);
        }

        public void CopyTo(TValue[] array, int arrayIndex)
        {
            this.CheckChanges();
            if (this.List != null)
            {
                this.List.CopyTo(array, arrayIndex);
            }
        }

        void ICollection.CopyTo(Array array, int index)
        {
            this.CheckChanges();
            if (this.List != null)
            {
                ((IList) this.List).CopyTo(array, index);
            }
        }

        public List<TValue>.Enumerator GetEnumerator()
        {
            this.CheckChanges();
            return (this.List ?? EmptyList).GetEnumerator();
        }

        public TValue this[int index]
        {
            get
            {
                this.CheckChanges();
                if (this.List == null)
                {
                    throw new ArgumentOutOfRangeException(nameof(index));
                }

                return this.List[index];
            }
            set
            {
                this.CheckChanges();
                if (this.List == null)
                {
                    throw new ArgumentOutOfRangeException(nameof(index));
                }

                this.List[index] = value;
            }
        }

        public int Count
        {
            get
            {
                this.CheckChanges();
                if (this.List == null)
                {
                    return 0;
                }

                return this.List.Count;
            }
        }

        // The following methods call the ones from above.

        bool IList.Contains(object value)
        {
            return this.Contains((TValue) value);
        }

        int IList.IndexOf(object value)
        {
            return this.IndexOf((TValue) value);
        }

        IEnumerator<TValue> IEnumerable<TValue>.GetEnumerator()
        {
            return this.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }

        int IList.Add(object value)
        {
            this.Add((TValue) value);
            return this.Count - 1;
        }

        void IList.Insert(int index, object value)
        {
            this.Insert(index, (TValue) value);
        }

        void IList.Remove(object value)
        {
            this.Remove((TValue) value);
        }

        object IList.this[int index]
        {
            get => this[index];
            set => this[index] = (TValue) value;
        }

        bool IList.IsReadOnly => false;

        bool ICollection<TValue>.IsReadOnly => false;

        bool IList.IsFixedSize => false;

        object ICollection.SyncRoot => throw new NotSupportedException();

        bool ICollection.IsSynchronized => false;

        public MultiDictionary<TKey, TValue> Parent { get; }

        public TKey Key { get; }

        private byte Version;
        internal List<TValue> List;
        private static readonly List<TValue> EmptyList = new List<TValue>();
    }
}
